### DISCUSSION TITLE
After RN/Xcode upgrade `FirebaseCoreInternal.modulemap' not found`

### DISCUSSION THREAD

**Discussion by jinhale** [invertase/react-native-firebase#7207] [Q&A] [answered]
**How can I restore a successful build process?**

After manually [upgrading][1] from RN version 0.70.2 to RN version 0.72.0 using a new project with code copied from my existing project, the new and existing project will no longer build using Xcode. I'm tying to build according to `Any iOS Device (arm64)`, which normally works. Here is the error:

`
~/Library/Developer/Xcode/DerivedData/<project-name>-aobgsqdmebtxwebkbxofzpscossc/Build/Intermediates.noindex/<project-name>.build/Debug-iphoneos/<project-name>.build/DerivedSources/<project-name>_vers.c module map file '~/Library/Developer/Xcode/DerivedData/<project-name>-aobgsqdmebtxwebkbxofzpscossc/Build/Products/Debug-iphoneos/FirebaseCoreInternal/FirebaseCoreInternal.modulemap' not found
`

When I look up modulemap files it seems to be related to manually linking files which is no longer applicable so I'm left confused. 

I have added `GoogleService-Info.plist` and checked to ensure the [configuration code][2] for FireBase was retained to the best of my ability. 

A similar [problem][3] has no accepted solution, but I should mention I have tried the obvious suggestions: close and reopen Xcode, clean before building, changing the minimum `iOS Deployment Target`, and removing `node_modules` and `Pods/` folder before reinstalling using `yarn` and `pod install`.

Here is `npx react-native info` output:

`
System:
  OS: macOS 13.4
  CPU: (4) x64 Intel(R) Core(TM) i3-8100B CPU @ 3.60GHz
  Memory: 183.78 MB / 8.00 GB
  Shell:
    version: 3.2.57
    path: /bin/bash
Binaries:
  Node:
    version: 20.3.1
    path: /usr/local/bin/node
  Yarn:
    version: 1.21.1
    path: ~/npm-global/bin/yarn
  npm:
    version: 9.6.7
    path: /usr/local/bin/npm
  Watchman:
    version: 2023.06.12.00
    path: /usr/local/bin/watchman
Managers:
  CocoaPods:
    version: 1.12.1
    path: ~/.rvm/gems/ruby-2.7.5/bin/pod
SDKs:
  iOS SDK:
    Platforms:
      - DriverKit 22.4
      - iOS 16.4
      - macOS 13.3
      - tvOS 16.4
      - watchOS 9.4
  Android SDK: Not Found
IDEs:
  Android Studio: Not Found
  Xcode:
    version: 14.3.1/14E300c
    path: /usr/bin/xcodebuild
Languages:
  Java: Not Found
  Ruby:
    version: 2.7.5
    path: ~/.rvm/rubies/ruby-2.7.5/bin/ruby
npmPackages:
  "@react-native-community/cli": Not Found
  react: Not Found
  react-native: Not Found
  react-native-macos: Not Found
npmGlobalPackages:
  "*react-native*": Not Found
Android:
  hermesEnabled: false
  newArchEnabled: false
iOS:
  hermesEnabled: true
  newArchEnabled: false
`

Here is a snippet of changes to `package.json`: 

`
"@react-native-firebase/analytics": [-"^16.4.0",-]{+"^18.1.0",+}
"@react-native-firebase/app": [-"^16.4.0",-]{+"^18.1.0",+}
"@react-native-firebase/messaging": [-"^16.4.0",-]{+"^18.1.0",+}
`

  [1]: https://react-native-community.github.io/upgrade-helper/?from=0.70.2&to=0.72.0
  [2]: https://rnfirebase.io/#configure-firebase-with-ios-credentials
  [3]: https://stackoverflow.com/questions/69319871/fastlane-deploy-ios-fatal-error-yogakit-modulemap-not-found

**Comment by mikehardy** [comment]
It is generally an intractable problem attempting to slowly tease sufficient information out of a project I can't see in order to figure out why the build won't work.

In place of attempting to do that, I have a fully scripted demonstrator repo that starts with `npx react-native init` and bit-by-bit adds in everything necessary to successfully build (assuming you have a plist / json file for google services)

I haven't officially promoted the react-native 0.72 version of it but I do testing with the react-native releases working group and I had it working as recently as react-native 0.72.0-rc05 - I don't anticipate anything big needs to change for the final release version: https://github.com/mikehardy/rnfbdemo/blob/rn72/make-demo.sh

Carefully read the comments and examine what the script does, maybe even try it as a scratch pad, it should work

I will note that if there is something related to FirebaseCoreInternal, please make sure you are using `use_frameworks! :linkage => :static` with no unsupported workarounds. FirebaseCoreInternal is a Swift module and it's linkage is problematic when the workarounds are attempted.

**Comment by anhquan99** [comment]
I found the solution at https://github.com/facebook/react-native/issues/29605#issuecomment-861783673

**Comment by jyotiprakash111** [comment]
Current RN Version 0.70, i am trying to upgrage from - 0.63 to 0.70, i am facing the above error, i have tried everything nothing is working for me as of now.

**Reply by jyotiprakash111** [reply]
System:
    OS: macOS 14.0
    CPU: (12) x64 Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
    Memory: 571.49 MB / 16.00 GB
    Shell: 3.2.57 - /bin/bash
  Binaries:
    Node: 18.18.0 - /usr/local/bin/node
    Yarn: Not Found
    npm: 9.8.1 - /usr/local/bin/npm
    Watchman: 2023.09.25.00 - /usr/local/bin/watchman
  Managers:
    CocoaPods: 1.12.1 - /usr/local/bin/pod
  SDKs:
    iOS SDK:
      Platforms: DriverKit 23.0, iOS 17.0, macOS 14.0, tvOS 17.0, watchOS 10.0
    Android SDK: Not Found
  IDEs:
    Android Studio: 2022.3 AI-223.8836.35.2231.10811636
    Xcode: 15.0.1/15A507 - /usr/bin/xcodebuild
  Languages:
    Java: 20.0.2 - /usr/bin/javac
  npmPackages:
    @react-native-community/cli: Not Found
    react: 18.2.0 => 18.2.0 
    react-native: ^0.70.0 => 0.70.14 
    react-native-macos: Not Found
  npmGlobalPackages:
    *react-native*: Not Found

**Reply by mikehardy** [reply]
Interesting, your cocoapod is out of date (1.14.2) is current but your Java is arguably too new (20? Does that even work? I think 17 is safer at the moment) and react-native 0.70 is old, why not 0.72.6 if you are upgrading things?

There is no problem with this repo though, and we don't have the resources to debug everyone's build problems since they are infinitely variable

Here is a script that works though, and will show you exactly how to do the minimal integration for a successful iOS build, on Xcode 15 etc https://github.com/mikehardy/rnfbdemo/blob/main/make-demo.sh

**Reply by GrkikyanRobert** [reply]
cool situation

**Comment by minhchienwikipedia** [comment]
@jinhale  I'm getting the same issue, do you have a solution?

**Reply by jinhale** [reply, marked as answer]
In the end I upgraded the OS, and XCode, then I created a new project adding each package one at a time. This worked, but it is a periodic issue with different versions.

**Reply by davidfcarr** [reply]
Can you be any more specific about the process you've followed? Is it necessary to recreate project.json and other assets by adding npm packages one at a time? Or can the "one at a time" process be limited to redoing the Podfile / pod install?

My situation is I inherited a project from another developer, cloned a repo from github and have overcome other dependency issues but now have hit this seemingly unsolvable modulemap roadblock. Trying to get a better understanding of how much effort it will take to make progress.

Do you understand what this modulemap file is and where it's supposed to come from? Is it supposed to be installed in the Pods directory or some other system directory?

**Reply by mikehardy** [reply]
In my experience if you ever see the modulemap error, you have a pretty low-level compilation issue
I've never see it on current versions with a correctly configured Podfile (that is, mostly stock from upstream and taking care to make sure `use_frameworks! :linkage => :static` is set, as firebase-ios-sdk requires)

You can read through the comments in the build script I use to demonstrate a from-scratch react-native + react-native-firebase project build here: https://github.com/mikehardy/rnfbdemo/blob/main/make-demo.sh

That project works from scratch every time with no modulemap errors, so may serve as a sort of sanity check on what a fresh super clean project should look like with regard to Podfile

**Reply by davidfcarr** [reply]
Honestly, I don't understand how your demo relates to issues such as Podfile configuration.

Can you provide an example of what a properly configured Podfile should look like. Where does the use_frameworks directive go?

**Reply by mikehardy** [reply]
> Honestly, I don't understand how your demo relates to issues such as Podfile configuration.

My demo generates a correctly configured Podfile, and it relates to you having build issues in exactly that way.
Run the demo, out pops a correctly configured Podfile

> Can you provide an example of what a properly configured Podfile should look like. Where does the use_frameworks directive go?

I'll do better! I'll give you a fully working demo project with correctly configured everything, but I'll only do it in a fully automated way so that I can maintain it going forward, and I do this...with my demo script :-)

**Reply by mikehardy** [reply]
also perhaps of interest, you may always consult our e2e test app for a look at how things *could* work. It is unique in some ways but does run so may serve as an example https://github.com/invertase/react-native-firebase/blob/9b5c405b2933c84daad561117a3eebacc65cbb7e/tests/ios/Podfile

**Reply by davidfcarr** [reply]
Thank you!

**Reply by davidfcarr** [reply]
I appreciate your guidance, but I'm still stuck on this error. Can I ask you to look at this Podfile and let me know if you can spot any errors?

`require_relative '../node_modules/react-native/scripts/react_native_pods'
require_relative '../node_modules/@react-native-community/cli-platform-ios/native_modules'

# Uncomment the next line to define a global platform for your project
platform :ios, '15.0'

target 'PrecintManager' do

# set this to static and toggle '$RNFirebaseAsStaticFramework' above to test static frameworks)
linkage = 'static' # ENV['USE_FRAMEWORKS']
if linkage != nil
  Pod::UI.puts "Configuring Pod with #{linkage}ally linked Frameworks".green
  use_frameworks! :linkage => linkage.to_sym
end

  config = use_native_modules!

  # Flags change depending on the env values.
  flags = get_default_flags()

  use_react_native!(
    :path => config[:reactNativePath],
    # Hermes is now enabled by default. Disable by setting this flag to false.
    # Upcoming versions of React Native may rely on get_default_flags(), but
    # we make it explicit here to aid in the React Native upgrade process.
    :hermes_enabled => flags[:hermes_enabled],
    :fabric_enabled => flags[:fabric_enabled],
    # Enables Flipper.
    #
    # Note that if you have use_frameworks! enabled, Flipper will not work and
    # you should disable the next line.
    # An absolute path to your application root.
    :app_path => "#{Pod::Config.instance.installation_root}/.."
  )

use_modular_headers!
  # Pods for PrecintManager
pod 'Firebase', :modular_headers => true
pod 'FirebaseCoreInternal', :modular_headers => true
pod 'GoogleUtilities', :modular_headers => true
pod 'FirebaseCore', :modular_headers => true

  target 'PrecintManagerTests' do
    inherit! :search_paths
    # Pods for testing
  end

end
`

This is what I got after trying to simplify the Podfile I inherited and make sure the static frameworks directive was included.

**Reply by mikehardy** [reply]
Absolutely no support will be forthcoming for "modular_header" workarounds ever. The need to use `use_frameworks! :linkage => :static` is clearly documented.

And people try to workaround it then wonder why they get build errors hmm 🤔 

Get rid of these and integrate things the way the docs specify

```
pod 'Firebase', :modular_headers => true
pod 'FirebaseCoreInternal', :modular_headers => true
pod 'GoogleUtilities', :modular_headers => true
pod 'FirebaseCore', :modular_headers => true
```

**Reply by davidfcarr** [reply]
I assume the previous developer got advice to use those headers commands. I've been through several more rounds of revising the Podfile, running pod integrate / pod install, and updating the Firebase NPM files in case that was the issue. Still no luck.

Here is the current Podfile, which tries to preserve the code I inherited but without those header references.

`require_relative '../node_modules/react-native/scripts/react_native_pods'
require_relative '../node_modules/@react-native-community/cli-platform-ios/native_modules'

platform :ios, '18.0'
prepare_react_native_project!

flipper_config = FlipperConfiguration.disabled

target 'PrecintManager' do
  config = use_native_modules!
  use_frameworks! :linkage => :static
  $RNFirebaseAsStaticFramework = true

  # Add other pods here
  pod 'GoogleUtilities'
  pod 'FirebaseCore'  
  pod 'FirebaseDatabase' 
  pod 'FirebaseMessaging'
  
  # Flags change depending on the env values.
  flags = get_default_flags()

  use_react_native!(
    :path => config[:reactNativePath],
    # Hermes is now enabled by default. Disable by setting this flag to false.
    # Upcoming versions of React Native may rely on get_default_flags(), but
    # we make it explicit here to aid in the React Native upgrade process.
    :hermes_enabled => flags[:hermes_enabled],
    :fabric_enabled => flags[:fabric_enabled],
    # Enables Flipper.
    #
    # Note that if you have use_frameworks! enabled, Flipper will not work and
    # you should disable the next line.
    :flipper_configuration => flipper_config,
    # An absolute path to your application root.
    :app_path => "#{Pod::Config.instance.installation_root}/.."
  )

  target 'PrecintManagerTests' do
    inherit! :complete
    # Pods for testing
  end

end`

I still get these module map errors
/Users/davidfcarr/Code/precinct-manager-frontend/ios/PrecintManager/main.m module map file '/Users/davidfcarr/Library/Developer/Xcode/DerivedData/PrecintManager-ecckizywzbrvaodvgiokpdwzxlgt/Build/Products/Debug-iphonesimulator/FirebaseCoreInternal/

/Users/davidfcarr/Library/Developer/Xcode/DerivedData/PrecintManager-ecckizywzbrvaodvgiokpdwzxlgt/Build/Intermediates.noindex/PrecintManager.build/Debug-iphonesimulator/PrecintManager.build/DerivedSources/PrecintManager_vers.c module map file '/Users/davidfcarr/Library/Developer/Xcode/DerivedData/PrecintManager-ecckizywzbrvaodvgiokpdwzxlgt/Build/Products/Debug-iphonesimulator/FirebaseCoreInternal/FirebaseCoreInternal.modulemap' not found

When I try digging into the directory referenced in these messages, I do see module.modulemap file in one of the subdirectories but apparently not what the compiler expects to find there.

![Module Map](https://github.com/user-attachments/assets/e44276f4-41e6-4225-a10c-32ffbd9edc61)

Is there any hope for getting this working?

**Reply by mikehardy** [reply]
It works for me literally every time. I have a fully automated build test that I use to troubleshoot problems like this.
It works. Every time.
https://github.com/mikehardy/rnfbdemo/blob/main/make-demo.sh

**Comment by kailash-sssg** [comment]
@jinhale I'm getting the same issue, do you have a solution? Did't find a way to fix it.
 having the upgraded OS and XCode which is OS- Sonoma, and XCode - 15.2

**Comment by rehan-taiftec** [comment]
any solution i am getting the same error after upgrading to the latest mac os

**Reply by kailash-sssg** [reply]
@rehan-taiftec, As of now only solution or workaround that worked for me is went through the way that has been mentioned by @jinhale here https://github.com/invertase/react-native-firebase/discussions/7207#discussioncomment-8346616

**Comment by minhchienwikipedia** [comment]
My solution is using `use_frameworks! :linkage => :static`

**Reply by vedant6565** [reply]
on which file write this please mention properly

**Reply by minhchienwikipedia** [reply]
on ios/Podfile

**Reply by muslimmuda15** [reply]
Please, Can you write full the code?

**Comment by SamDuke99** [comment]
This solution works for me:

1. Go into build settings for your project
2. Head to architectures
3. Open up excluded architectures
4. Exclude 'arm64'

See pic for more context:
![image](https://github.com/invertase/react-native-firebase/assets/58332516/9c1ae1d9-af58-4acf-b787-d8ac436d2ce1)

**Reply by honguyen25121991** [reply]
Thank you ! It's work for me

**Reply by moin-nimblechapps** [reply]
Thank you ! This work for me.

**Reply by ssthil** [reply]
Thanks, @SamDuke99.
It's worked for me as well.

Note: In addition, I have faced `Flipper` issues as attached.
![image](https://github.com/user-attachments/assets/7253bbe6-1977-4756-9415-e109ba1b67e5)

However, I have followed the other thread (https://github.com/facebook/react-native/issues/43335) to fix it.

🚀 👍  **I've followed the fix below, and it worked fine for me.** ❤️ 

![image](https://github.com/user-attachments/assets/b3dbc269-6059-41e6-bb3a-391ec4d35b11)

Project Info:
`React Native Version: 0.62.2`

> System:

    OS: macOS 15.1.1
    CPU: (12) arm64 Apple M2 Pro
    Memory: 95.83 MB / 32.00 GB
    Shell: 5.9 - /bin/zsh
    
  Binaries:
  
    Node: 16.20.2
    Yarn: 1.22.17
    npm: 8.19.4 
    Watchman: 2024.10.28.00 
  Managers:
  
    CocoaPods: 1.16.2 
    
  SDKs:
    iOS SDK:
      Platforms: DriverKit 24.1, iOS 18.1, macOS 15.1

**Reply by mikehardy** [reply]
it is unreasonable to expect react-native 0.62 to work in any context I can think of.
Flipper is completely dead in current react-native and no time invested in making it work is a good investment
I'd get your project upgraded to react-native 0.76.3

**Reply by ssthil** [reply]
@mikehardy Thanks for your reply, and I apologize if my previous response was unclear. I have **resolved** the issue (working fine) and shared the version of `React Native` I used for future reference in case anyone encounters the same problem.

**Reply by BThanhNhut** [reply]
Thanks, it's work for me

**Comment by AteebNoOne** [comment]
/Users/mac/Library/Developer/Xcode/DerivedData/CI_Builders-doiyvzlocwxpabcvsqshiqmtknxx/Build/Intermediates.noindex/CI_Builders.build/Debug-iphonesimulator/CI_Builders.build/DerivedSources/CI_Builders_vers.c module map file '/Users/mac/Library/Developer/Xcode/DerivedData/CI_Builders-doiyvzlocwxpabcvsqshiqmtknxx/Build/Products/Debug-iphonesimulator/FirebaseCoreInternal/FirebaseCoreInternal.modulemap' not found


React native info:

System:
  OS: macOS 13.6.6
  CPU: (4) x64 Intel(R) Core(TM) i5-7267U CPU @ 3.10GHz
  Memory: 701.54 MB / 16.00 GB
  Shell:
    version: "5.9"
    path: /bin/zsh
Binaries:
  Node:
    version: 21.7.2
    path: /usr/local/bin/node
  Yarn: Not Found
  npm:
    version: 10.5.0
    path: /usr/local/bin/npm
  Watchman: Not Found
Managers:
  CocoaPods:
    version: 1.15.2
    path: /usr/local/bin/pod
SDKs:
  iOS SDK:
    Platforms:
      - DriverKit 23.2
      - iOS 17.2
      - macOS 14.2
      - tvOS 17.2
      - visionOS 1.0
      - watchOS 10.2
  Android SDK: Not Found
IDEs:
  Android Studio: Not Found
  Xcode:
    version: 15.2/15C500b
    path: /usr/bin/xcodebuild
Languages:
  Java: Not Found
  Ruby:
    version: 3.3.0
    path: /usr/local/bin/ruby
npmPackages:
  "@react-native-community/cli": Not Found
  react:
    installed: 18.2.0
    wanted: 18.2.0
  react-native:
    installed: 0.73.6
    wanted: 0.73.6
  react-native-macos: Not Found
npmGlobalPackages:
  "*react-native*": Not Found
Android:
  hermesEnabled: true
  newArchEnabled: false
iOS:
  hermesEnabled: true
  newArchEnabled: false
  
  
  Podfile:
  
  
require Pod::Executable.execute_command('node', ['-p',
  'require.resolve(
    "react-native/scripts/react_native_pods.rb",
    {paths: [process.argv[1]]},
  )', __dir__]).strip

platform :ios, min_ios_version_supported


prepare_react_native_project!

# If you are using a `react-native-flipper` your iOS build will fail when `NO_FLIPPER=1` is set.
# because `react-native-flipper` depends on (FlipperKit,...) that will be excluded
#
# To fix this you can also exclude `react-native-flipper` using a `react-native.config.js`
# ```js
# module.exports = {
#   dependencies: {
#     ...(process.env.NO_FLIPPER ? { 'react-native-flipper': { platforms: { ios: null } } } : {}),
# ```
flipper_config = ENV['NO_FLIPPER'] == "1" ? FlipperConfiguration.disabled : FlipperConfiguration.enabled

linkage = ENV['USE_FRAMEWORKS']
if linkage != nil
  Pod::UI.puts "Configuring Pod with #{linkage}ally linked Frameworks".green
  use_frameworks! :linkage => linkage.to_sym
end


target 'CI_Builders' do
  config = use_native_modules!
  pod 'Firebase', :modular_headers => true
  pod 'FirebaseCoreInternal', :modular_headers => true
  pod 'GoogleUtilities', :modular_headers => true
  use_react_native!(
    :path => config[:reactNativePath],
    # Enables Flipper.
    #
    # Note that if you have use_frameworks! enabled, Flipper will not work and
    # you should disable the next line.
    :flipper_configuration => flipper_config,
    # An absolute path to your application root.
    :app_path => "#{Pod::Config.instance.installation_root}/.."
  )

  target 'CI_BuildersTests' do
    inherit! :complete
    # Pods for testing
  end

  post_install do |installer|
    # https://github.com/facebook/react-native/blob/main/packages/react-native/scripts/react_native_pods.rb#L197-L202
    react_native_post_install(
      installer,
      config[:reactNativePath],
      :mac_catalyst_enabled => false
    )
  end
end



# is there any solution for it?

**Comment by bilal-prog** [comment]
I get that error when I add tvOS to the targets, PLZ is there any solution ?

**Comment by tigerjiang** [comment]
I get that error when I add tvOS to the targets, PLZ is there any solution ?

**Comment by Kayskip** [comment]
This solution worked for me.

Workspace -> app -> Build Settings -> Architectures
then
Excluded Architectures -> Debug -> Any iOS Simulator -> Make Blank/Delete (It was previously Any SDK)

Using Xcode project version 12.0
